src/driver/kvm.rs

```diff
@@ -106,13 +132,82 @@ impl Introspectable for Kvm {
         Ok(())
     }
 
+    fn toggle_intercept(
+        &mut self,
+        vcpu: u16,
+        intercept_type: InterceptType,
+        enabled: bool,
+    ) -> Result<(), Box<dyn Error>> {
+        match intercept_type {
+            InterceptType::Cr(micro_cr_type) => {
+                let kvmi_cr = match micro_cr_type {
+                    CrType::Cr0 => KVMiCr::Cr0,
+                    CrType::Cr3 => KVMiCr::Cr3,
+                    CrType::Cr4 => KVMiCr::Cr4,
+                };
+                Ok(self.kvmi.control_cr(vcpu, kvmi_cr, enabled)?)
+            }
+        }
+    }
+
+    fn listen(&mut self, timeout: u32) -> Result<Option<Event>, Box<dyn Error>> {
+        // wait for next event and pop it
+        debug!("wait for next event");
+        if self.kvmi.wait_event(timeout.try_into().unwrap())?.is_none() {
```

Comment:
nitpick: `kvmi.wait_event` returns `None` or an empty `Some()`. I would not expect Some() to yield no value. Maybe use a custom enum event/no_event instead?

Reply:
Also, I don't see why `wait_event` shouldn't return the event directly. Is there a special reason for not popping the event directly in `wait_event`?

Reply:
I implemented the KVMi Rust API by doing a 1:1 mapping with the libkvmi functions.
But I suppose we could wrap these 2 calls into a single one `pop_event()` function.

Reply:
Either way works. :+1: